elixir/apps/domain/lib/domain/replication/connection.ex

```diff
@@ -425,67 +437,57 @@ defmodule Domain.Replication.Connection do
           columns: columns
         }
 
-        {:noreply, ack(server_wal_end),
-         %{state | relations: Map.put(state.relations, id, relation)}}
+        {:noreply, %{state | relations: Map.put(state.relations, id, relation)}}
       end
 
-      defp handle_message(%Decoder.Messages.Insert{} = msg, server_wal_end, state) do
-        unless state.error_threshold_exceeded? do
-          {op, table, _old_data, data} = transform(msg, state.relations)
-          :ok = on_insert(server_wal_end, table, data)
-        end
+      defp handle_write(%Decoder.Messages.Insert{} = msg, server_wal_end, state) do
+        process_write(msg, server_wal_end, state)
+      end
 
-        {:noreply, ack(server_wal_end), state}
+      defp handle_write(%Decoder.Messages.Update{} = msg, server_wal_end, state) do
+        process_write(msg, server_wal_end, state)
       end
 
-      defp handle_message(%Decoder.Messages.Update{} = msg, server_wal_end, state) do
-        unless state.error_threshold_exceeded? do
-          {op, table, old_data, data} = transform(msg, state.relations)
-          :ok = on_update(server_wal_end, table, old_data, data)
-        end
+      defp handle_write(%Decoder.Messages.Delete{} = msg, server_wal_end, state) do
+        process_write(msg, server_wal_end, state)
+      end
```

Comment:
I may be overlooking something, but these are all doing the exact same thing yes?  Was there a reason to split them out in to separate function definitions?

Reply:
Well it's pattern matching on just the `Insert`, `Update`, `Delete` messages.

Reply:
We don't want to call process_write for the other message types.

Reply:
Oh, is the `handle_write` on line 443 a catch all for all other types of messages?  I was thinking line 443 was the only one that was specific and these three were essentially the "all other messages call process_write".

I guess what I'm really asking is, are these the only message types that will ever get passed?
```
%Decoder.Messages.Relation{}
%Decoder.Messages.Insert{}
%Decoder.Messages.Update{}
%Decoder.Messages.Delete{}
```

Reply:
No those are just the ones we update state for. Starting on line  There's a few more here: https://github.com/firezone/firezone/pull/9733/files?new_files_changed=true#diff-ce1fe0c661c1ff198712e57eb249878194786ce317f2b4b920fcd501a1cf9419R519

Reply:
That link seems to just take me right back to this comment for some reason.  Although I suppose it doesn't matter too much, because my next question is, if those aren't the only messages that will be passed in to these functions won't you need a default "no-op" version of this function for all of the other types of messages?  Wouldn't we crash due to not having a matching function signature?   This is what ultimately led me to ask my initial question.  It felt like there was on special case `%Decoder.Messages.Relation` and all of the others were doing the same thing.

Reply:
Starting on line 513 in that are the other messages. The `Decoder` module creates an `Unsupported` msg type that we explicitly match and log a warning for down there.

We could collapse a few of the ones we don't care about but I enumerated them just for documentation / clarity.

Reply:
See https://github.com/firezone/firezone/blob/main/elixir/apps/domain/lib/domain/replication/decoder.ex

Reply:
I.e. - we are matching on each message type so there is no reason for a catch-all. The handle_write function is a `defp` and not `@impl true` so it's by us elsewhere in this module. This hasn't changed since the introduction of the replication connection module, only shortened a bit here in this PR.

Reply:
Ah, ok, I see now.  I'm realizing I missed that this was part of a macro.